test/rule_definition_tools.jl

```diff
@@ -0,0 +1,65 @@
+@testset "rule_definition_tools.jl" begin
```

Comment:
It looks like there's quite a lot of repeated code here. Did you consider writing a function to test that something has been successfully "non_differentiable"d?

Reply:
Repeating code mades it read straight forward, and each is different enough that abstracting the tests would be make them harder to read.

Reply:
It's not clear to me that that is true. Something like the following ought to do the majority of the work:
```julia
function test_nondifferentiable(foo, args, dargs, dy)
    @test frule(dargs, foo, args...) == foo(args...)

    y, pb = rrule(foo, args...)
    @test y == foo(args...)
    @test pb(dy) == (Zero(), map(_ -> DoesNotExist(), args)...)
end
```

To my mind this is more readable.

I'm not going to object to merging this over this though -- I'm happy to stick with what you've done if you feel strongly that it's more readable.